resource/traversers/dfu_flexible.cpp

```diff
@@ -0,0 +1,365 @@
+                            std::unordered_map<resource_type_t, int64_t> &lowest)
+{
+    int rc = -1;
+    if (m_match->is_pruning_type (subsystem, type)) {
+        if (lowest.find (type) == lowest.end ())
+            lowest[type] = counts;
+        else if (lowest[type] > counts)
+            lowest[type] = counts;
+        rc = 0;
+    }
+    return rc;
+}
+
+void dfu_flexible_t::prime_jobspec (std::vector<Resource> &resources,
+                                    std::unordered_map<resource_type_t, int64_t> &to_parent)
+{
+    subsystem_t subsystem = m_match->dom_subsystem ();
+    for (auto &resource : resources) {
+        // If the resource is requested as exclusive in the
+        // jobspec, add it to the matcher's exclusive resource
+        // set. This ensures that the full resource set (which
+        // includes shadow resources) is emitted.
+        if (resource.exclusive == Jobspec::tristate_t::TRUE)
+            m_match->add_exclusive_resource_type (resource.type);
+        // Use minimum requirement because you don't want to prune search
+        // as far as a subtree satisfies the minimum requirement
+        accum_if (subsystem, resource.type, resource.count.min, to_parent);
+        prime_jobspec (resource.with, resource.user_data);
+
+        // Or slots should use a minimum of values rather than an accumulation
+        // otherwise possible matches may be filtered out
+        if (resource.type == slot_rt) {
+            for (auto &aggregate : resource.user_data) {
+                min_if (subsystem,
+                        aggregate.first,
+                        resource.count.min * aggregate.second,
+                        to_parent);
+            }
+        } else {
+            for (auto &aggregate : resource.user_data) {
+                accum_if (subsystem,
+                          aggregate.first,
+                          resource.count.min * aggregate.second,
+                          to_parent);
+            }
+        }
+    }
+}
+
+std::tuple<std::map<resource_type_t, int>, int, int> dfu_flexible_t::select_or_config (
+    const std::vector<Resource> &slots,
+    std::map<resource_type_t, int> resource_counts,
+    unsigned int nslots,
+    std::unordered_map<Key, std::tuple<std::map<resource_type_t, int>, int, int>, Hash> &or_config)
+{
+    int best = -1;
+    int i = -1;
+    Key index = Key (resource_counts);
+
+    // if available, use precomputed result
+    auto it = or_config.find (resource_counts);
+    if (it != or_config.end ())
+        return it->second;
+
+    // if there is only a single slot, no need to compute
+    if (slots.size () == 1) {
+        or_config[index] = std::make_tuple (resource_counts, nslots, 0);
+        return or_config[index];
+    }
+
+    for (auto slot : slots) {
+        int test;
+        ++i;
+        bool match = true;
+        std::map<resource_type_t, int> updated_counts;
+        updated_counts = resource_counts;
```

Comment:
Possibly just a style suggestion, but you can combine these two lines:

```suggestion
        std::map<resource_type_t, int> updated_counts = resource_counts;
```